### PR TITLE
ng/jupyter: Pipe stdout to output buffer

### PR DESCRIPTION
**Pull request by ksshannon**
Create a pipe that temporarily pipes stdout to the session buffer so
fmt.Print*() and friends print to jupyter, not stdout.

I think users would expect those funcs to get printed out, but I'm not sure if this is expected behavior.  Also I'm not sure if I implemented it correctly.  Writing to buf could be racy.

Updates #16

### CODE REVIEW

File: jupyter/jupyter.go [hunk 1]
```diff
@@ -299,6 +293,34 @@ func (s *server) execute(c *conn, req *message) error {
...
+	session.Stdout = buf
+	session.Stderr = buf
+
+	done := make(chan struct{})
+	go func() {
+		io.Copy(buf, r)
+		close(done)
+	}()
+	defer func() {
+		w.Close()
```
Comment (crawshaw) [inline jupyter/jupyter.go]: The timing is off on this. You close the pipe here (when the function exits), and when <-done returns it is safe to read from `buf`. But `buf.String()` is called long before the defer as an argument to publishIO.

I would prefer this code was inside the Exec method in `ngcore`.

(Long-term I think it should go a little deeper, and we should feed the evaluator different stdout/stderr global variables per session, but this pipe trick in ngcore is a good start.)

Comment (ksshannon) [reply jupyter/jupyter.go]: Ah, I see.  Thanks for the look.

**Comment by ksshannon** [comment]
Closing in favor of a better strategy.

**Comment by crawshaw** [comment]
@ksshannon do you plan on sending another PR for using a Pipe inside `ngcore`? (I think the extra plumbing I mentioned above can wait for another time.) If not, I'll file an issue so we remember.

Also, thank you for sending a PR!

**Comment by ksshannon** [comment]
@crawshaw, I'll take a crack at it.  If I get hung up or can't do it, I can file an issue.  Thanks for the feedback.

**Comment by crawshaw** [comment]
I've been thinking a bit more about this, and I think we should change the ngcore package interface a bit.

Right now it has fields for Stdout/Stderr that are io.Writers, and it adapts a pipe to them as needed for passing into job executions. Instead, we should make Stdin/Stdout/Stderr `*os.File` objects in the interface. Let callers of the ngcore package create a pipe and set it as they want.

Then you can take those *os.File objects, and use them to temporarily set os.Stdin, Stdout, Stderr during an eval execution.

**Comment by ksshannon** [comment]
@crawshaw #104 is my new attempt.  I added the *os.File changes mentioned.  I'm not sure about the ioutil.Discard->os.Create(os.DevNull) change, there is probably a cleaner way to do that.

**Pull request closed without merging**